### PR TITLE
strutil: introducing MatchCounter

### PR DESCRIPTION
**Pull request by chipaca**
A MatchCounter is a discarding io.Writer that retains up to N
matches to its Regexp before just counting matches.

### CODE REVIEW

File: strutil/matchcounter.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	N int
+
+	count   int
+	matches []string
+	partial []byte
+}
+
+func (w *MatchCounter) Write(p []byte) (int, error) {
+	n := len(p)
+	if len(w.partial) > 0 {
```
Comment (zyga) [inline strutil/matchcounter.go]: Nitpick: would this work without the conditional check on `len(w.partial)`, I mean, the idx would just be -1.

Comment (chipaca) [reply strutil/matchcounter.go]: Can you ask that again, but slower? It's late...

File: strutil/matchcounter_test.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+func (mcSuite) TestMatchCounterPartials(c *check.C) {
+	// now we know the whole thing matches expected, we check partials
+	buf := []byte(out)
+	expected := []string{
+		"Failed to write /tmp/1/modules/4.4.0-112-generic/modules.symbols, skipping",
+		"Write on output file failed because No space left on device",
+		"writer: failed to write data block 0",
+	}
+
+	for step := 1; step < 100; step++ {
```
Comment (mvo5) [inline strutil/matchcounter_test.go]: Nice test!

File: strutil/matchcounter.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		return n, nil
+	}
+	idx++
+	w.partial = p[idx:]
+	w.check(p[:idx])
+	return n, nil
+}
+
+func (w *MatchCounter) check(p []byte) {
+	matches := w.Regexp.FindAll(p, -1)
```
Comment (bboozzoo) [inline strutil/matchcounter.go]: No need to find all matches if the code will discard anything more than N:
```
matches := w.Regexp.FindAll(p, w.N)
```

Comment (chipaca) [reply strutil/matchcounter.go]: but we don't just discard them; we count them.
And as it's `[][]byte`, I assume it's just a reslicing of `p`.

Comment (bboozzoo) [reply strutil/matchcounter.go]: Oh, right, they're still counted but only N are kept around for later. Fair enough.

**Pull request merged**